### DISCUSSION TITLE
Wrong Preview URL when selecting Google Cloud Storage (GCS) as the disk

### DISCUSSION THREAD

**Discussion by nidhalkratos** [livewire/livewire#3483] [7. Bugs (archived)]
Hi, 
First of all, great job you guys are doing at the livewire project.

We are facing this issue and we believe it is actually a bug in the framework, When I change the disk config from null to 'gcs', the temporary image preview does not work.
(We need to use GCS instead of local storage because when having multiple machines serving the sites, requests will not necessarily go to the same machine and thus the image preview must be stored somewhere else other than the machine itself, hence GCS.

This is a fraction of the config/livewire.php file
```php
'temporary_file_upload' => [
        'disk' => 'gcs',        // Example: 'local', 's3'              Default: 'default'
        'rules' => null,       // Example: ['file', 'mimes:png,jpg']  Default: ['required', 'file', 'max:12288'] (12MB)
        'directory' => null,   // Example: 'tmp'                      Default  'livewire-tmp'
        'middleware' => null,  // Example: 'throttle:5,1'             Default: 'throttle:60,1'
        'preview_mimes' => [   // Supported file types for temporary pre-signed file URLs.
            'png', 'jpg', 'gif', 'bmp', 'svg', 'wav', 'mp4',
            'mov', 'avi', 'wmv', 'mp3', 'm4a',
            'jpeg', 'mpga', 'webp', 'wma',
        ],
    ],
```
And this is a fraction from the config/filesystems.php file
```php
'disks' => [

        'local' => [
            'driver' => 'local',
            'root' => storage_path('app'),
        ],

        'public' => [
            'driver' => 'local',
            'root' => storage_path('app/public'),
            'url' => env('APP_URL').'/storage',
            'visibility' => 'public',
        ],

        's3' => [
            'driver' => 's3',
            'key' => env('AWS_ACCESS_KEY_ID'),
            'secret' => env('AWS_SECRET_ACCESS_KEY'),
            'region' => env('AWS_DEFAULT_REGION'),
            'bucket' => env('AWS_BUCKET'),
            'url' => env('AWS_URL'),
            'endpoint' => env('AWS_ENDPOINT'),
        ],

        'gcs' => [
            'driver' => 'gcs',
            'project_id' => env('GOOGLE_CLOUD_PROJECT_ID', 'default-project-id'),
            'key_file' => [
                'type' => env('GOOGLE_CLOUD_ACCOUNT_TYPE'),
                'private_key_id' => env('GOOGLE_CLOUD_PRIVATE_KEY_ID'),
                'private_key' => env('GOOGLE_CLOUD_PRIVATE_KEY'),
                'client_email' => env('GOOGLE_CLOUD_CLIENT_EMAIL'),
                'client_id' => env('GOOGLE_CLOUD_CLIENT_ID'),
                'auth_uri' => env('GOOGLE_CLOUD_AUTH_URI'),
                'token_uri' => env('GOOGLE_CLOUD_TOKEN_URI'),
                'auth_provider_x509_cert_url' => env('GOOGLE_CLOUD_AUTH_PROVIDER_CERT_URL'),
                'client_x509_cert_url' => env('GOOGLE_CLOUD_CLIENT_CERT_URL'),
            ],
            'bucket' => env('GOOGLE_CLOUD_STORAGE_BUCKET', 'default-bucket'),
            'path_prefix' => env('GOOGLE_CLOUD_STORAGE_PATH_PREFIX', null), // optional: /default/path/to/apply/in/bucket
            'storage_api_uri' => env('GOOGLE_CLOUD_STORAGE_API_URI', null), // see: Public URLs below
            'visibility' => 'public', // optional: public|private
        ],

    ],
```
The image preview image url that livewire is generating is 
```
SOMEWEBSITE.COM/livewire/preview-file/SOMEFILE.JPG?expires=SOMETIMESTAMP&signature=SOMESIGNATURE
```
While it should be
```
static.SOMEWEBSITE.COM/livewire-tmp/SOMEFILE.JPG?expires=SOMETIMESTAMP&signature=SOMESIGNATURE
```

Note that when I try to access the second url directly it works, which means the image is uploaded correctly to gcs, the problem is the that livewire is not returning the correct url to that image so that the front end preview loads it.

**Comment by joshhanley** [comment]
@nidhalkratos have you configured your `.ENV` file to have a key `GOOGLE_CLOUD_STORAGE_API_URI=` that points to `static.SOMEWEBSITE.com`?

**Reply by nidhalkratos** [reply]
@joshhanley Yes, the variable GOOGLE_CLOUD_STORAGE_API_URI points to https://static.SOMEWEBSITE.com/, in fact, we are using it outside of livewire in an Image, Video, and Attachement models and it works fine. 

Here is a fraction of the Image model class
```php
protected $IMAGE_DIRECTORY_PATH = 'images/';
public function path()
{
    return $this->IMAGE_DIRECTORY_PATH . $this->{$this->selectedResolution};
}
public function url()
{
    return Storage::disk('gcs')->url($this->path());
}
```

And this is the version of livewire being used in the composer lock file `"livewire/livewire": "v2.4.4",`

**Reply by nidhalkratos** [reply]
So whenever we want to access any image link, it would look something like this:
```php
Image::find(SOMEID)->url() ;
// Or we set a specific image size first
Image::find(SOMEID)->large()->url();
```

**Comment by nidhalkratos** [comment]
I there anything we can do to fix this issue?
Can we mark It as a bug to get the community to have a look at it?

**Reply by joshhanley** [reply]
@nidhalkratos a PR that fixes this issue would be welcome!

You've already marked this as a bug by putting it in the bugs category 😁 but in terms of someone looking at it, you are probably in the best position to work out what is going on, as you already have google cloud setup and running.

You will need to dig in and make sure your storage driver is returning the correct URL (before Livewire touches it) and if it is, try and narrow down where Livewire might be modifying that.

The area in Livewire's code base you will need to investigate will be here.

https://github.com/livewire/livewire/blob/db71d14de36dde11b224183c57bdd913f7cb6110/src/TemporaryUploadedFile.php#L70-L73

Hope this helps! Let me know if you need any further guidance 🙂

**Reply by nidhalkratos** [reply]
Thank you @joshhanley very much, your guidance was helpful, found the bug, we needed a `isUsingGCS()` just like `isUsingS3` method to detect GCS. Now it works fine,  I submitted a pull request, hope it goes through so that we kick off the deployment cycle to update the site on the production cluster.

Also, we needed to upadate all of these lines 
```php
$url = $this->somelivewireattribute->getRealPath();
```
becomes 
```php
$url = $this->somelivewireattribute->temporaryUrl();
```
in all of our codebase to ensure we reference the right image at `https://static.website.com/livewire-tmp/blabla.jpg`

**Reply by joshhanley** [reply]
@nidhalkratos no worres! 🙂 glad you found it and thanks for the PR!
